Non english months in pdb headers 4449

exs-gbartlett opened the pull request:
<!--- Please read each of the following items and confirm by replacing
 !--the [ ] with a [X] --->

- [X] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [X] I have read the ``CONTRIBUTING.rst`` file, have run ``pre-commit``
locally, and understand that continuous integration checks will be used to
confirm the Biopython unit tests and style checks pass with these changes.

- [] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

<!--- Briefly describe the changes included in this pull request below
 !--- starting with 'Closes #...' if appropriate --->

Closes #4449 

Non-recognized months in pdb header no longer raise ValueError but return '0', the same behaviour as a month of 'xxx'. Unit test included.

peterjc commented:
Currently this silently ignores the bad month in a PDB file, using zero instead.

Unless our PDB team think otherwise, I would prefer this to consider the permissive/strict mode, and adjust the behavior accordingly.

CC @JoaoRodrigues @etal @jgreener64 (not an exhaustive list)

jgreener64 commented:
I don't have strong opinions here. On the one hand I don't think returning a zero will cause problems in user code. On the other hand it does say at https://www.wwpdb.org/documentation/file-format-content/format33/sect2.html#HEADER that

> The verification program checks that the deposition date is a legitimate date

so I would guess that all PDB entries have a legitimate date (I haven't checked).

peterjc commented:
I strongly suspect the test case here is from a third party tool exporting in PDB format, not a file from an official repository/mirror.

exs-gbartlett commented:
The test case is from a PDB file generated by a third party tool and is not a public repository file. Does biopython only support PDB files that have been deposited at wwPDB? The month 'xxx' is tolerated in the code without strict/permissive checks.

peterjc commented:
We try to parse "unofficial" files, but where they break the specification it is reasonable to give a warning or error.

In this case I personally would not want this to be silently parsed in strict mode.

exs-gbartlett commented:
Updated with warning and permissive flag carried through to the `_format_date` function. Updated unit tests. Hopefully this is an acceptable solution.

mdehoon commented on Bio/PDB/parse_pdb_header.py:
```diff
@@ -77,7 +88,17 @@ def _format_date(pdb_date):
         "Dec",
     ]
-    month = str(all_months.index(pdb_date[3:6]))
+    month_name = pdb_date[3:6]
```
```python
    try:
        month = "%02d" % all_months.index(month_name)
    except ValueError:
        if not permissive:
            raise ValueError(f"Non-standard month supplied: {month_name}.") from None
        month = "00"
```

exs-gbartlett replied:
That works - except we still need to raise a warning if it is permissive.

peterjc replied:
You could, or decide in permissive mode just treat the month as zero. That's a judgement call - and right now your files are the only example of this I'm aware of.

exs-gbartlett replied:
As per your comment 'where they break the specification it is reasonable to give a warning or error.' - I have put a warning in where they break the spec in permissive mode.

peterjc reviewed (changes requested):
A minor improvement aside (warning message and class), this looks good to me.

mdehoon commented on Bio/PDB/parse_pdb_header.py:
```diff
@@ -77,7 +87,22 @@ def _format_date(pdb_date):
+                f"Non-standard month in PDB header: {month_name}."
+            ) from None
+
+        warnings.warn(
```
I still object against warnings, as by default they get printed only once in one Python session.

peterjc commented:
This needs rebasing to address the merge conflicts.
